source/common/request_source_plugin_impl.cc

```diff
@@ -0,0 +1,118 @@
+#include "common/request_source_plugin_impl.h"
+
+#include "external/envoy/source/common/protobuf/message_validator_impl.h"
+#include "external/envoy/source/common/protobuf/utility.h"
+#include "external/envoy/source/exe/platform_impl.h"
+
+#include "api/client/options.pb.h"
+
+#include "common/request_impl.h"
+#include "common/request_source_impl.h"
+
+namespace Nighthawk {
+
+
+std::string DummyRequestSourceConfigFactory::name() const {
+  return "nighthawk.dummy-request-source-plugin";
+}
+
+Envoy::ProtobufTypes::MessagePtr DummyRequestSourceConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<nighthawk::request_source::DummyPluginRequestSourceConfig>();
+}
+
+RequestSourcePtr
+DummyRequestSourceConfigFactory::createRequestSourcePlugin(const Envoy::Protobuf::Message& message,
+                                                           Envoy::Api::Api&) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::request_source::DummyPluginRequestSourceConfig config;
+  Envoy::MessageUtil::unpackTo(any, config);
+  return std::make_unique<DummyRequestSourcePlugin>(config);
+}
+
+REGISTER_FACTORY(DummyRequestSourceConfigFactory, RequestSourcePluginConfigFactory);
+
+DummyRequestSourcePlugin::DummyRequestSourcePlugin(
+    const nighthawk::request_source::DummyPluginRequestSourceConfig& config)
+    : dummy_value_{config.has_dummy_value()
+                                                 ? config.dummy_value().value()
+                                                 : std::numeric_limits<double>::infinity()} {}
+RequestGenerator DummyRequestSourcePlugin::get() {
+  RequestGenerator request_generator = []() {
+    Envoy::Http::RequestHeaderMapPtr header = Envoy::Http::RequestHeaderMapImpl::create();
+    auto returned_request_impl = std::make_unique<RequestImpl>(std::move(header));
+    return returned_request_impl;
+  };
+  return request_generator;
+}
+
+void DummyRequestSourcePlugin::initOnThread() {}
+
+std::string FileBasedRequestSourceConfigFactory::name() const {
+  return "nighthawk.file-based-request-source-plugin";
+}
+
+Envoy::ProtobufTypes::MessagePtr FileBasedRequestSourceConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<nighthawk::request_source::FileBasedPluginRequestSourceConfig>();
+}
+
+RequestSourcePtr FileBasedRequestSourceConfigFactory::createRequestSourcePlugin(
+    const Envoy::Protobuf::Message& message, Envoy::Api::Api& api) {
+  const auto& any = dynamic_cast<const Envoy::ProtobufWkt::Any&>(message);
+  nighthawk::request_source::FileBasedPluginRequestSourceConfig config;
+  Envoy::MessageUtil util;
+  util.unpackTo(any, config);
+  RELEASE_ASSERT(api.fileSystem().fileSize(config.file_path()) < config.max_file_size().value(),
+                 "file size must be less than max_file_size");
+  auto temp_list = std::make_unique<nighthawk::client::RequestOptionsList>();
+  {
+    Envoy::Thread::LockGuard lock_guard(file_lock_);
```

Comment:
Why do we need to lock here? Can we spend a code-level comment on that?  Also, should we then also have a simple spam test for this, calling it from multiple threads? We might not even want to have any specific expectations in it, but just so we give tsan a good opportunity to have a say about this?
(sorry, I don't want to bloat this PR, but given that it's targetting production and not test code, this might be worth considering?)

Reply:
I added a test with multiple calls to create, though I'm not sure that this tests for what you want to test for. I added a comment about the lock also.

Reply:
Thanks for that comment, that is helpful. So now that the intent is clear, looking at the test you added, how about adding something like [1] to ensure multiple *concurrent* calls will only ever result in a single config file being read from disk? Such an amendment would also give the CI TSAN job a chance to verify.

```c++
void spamCall(std::function<void()>&& call_to_spam, const uint32_t num_threads) {
  std::vector<std::thread> threads(num_threads);
  auto barrier = std::make_unique<absl::Barrier>(num_threads);

  for (auto& thread : threads) {
    thread = std::thread([&call_to_spam, &barrier] {
      // Allow threads to accrue, to maximize concurrency on the call we are testing.
      if (barrier->Block()) {
        barrier.reset();
      }
      call_to_spam();
    });
  }
  for (std::thread& thread : threads) {
    thread.join();
  }
}

....

TEST_F(FileBasedRequestSourcePluginTest, CreateMultipleRequestSourcePluginReadsFileOnce) {
   constexpr uint32_t kNumThreads = 100;
   spamCall([kNumThreads]() { 
       .. insert the current test body which ensures a single config file is read that you have here...
    }, 
    kNumThreads);
}

```

Reply:
Also, it seems to me that there might be value to put every known implementation of `::initOnThread` through a similar test to help ensure concurrent calls of these stay well behaved during future maintenance. I don't think this is an issue now or with the code here, but if you agree on the principle, maybe file an issue on it and tag it as tech-debt?

Reply:
After further thought, this doesn't seem appropriate to have as a unit test. The primary problem being that if it were to be an appropriate unit test for a threading issue, then it is inherently trying to be non deterministic, which is not what a unit test should be. It seems like what we want is maybe closer to an integration test, but that's probably best done in a separate pr.

Reply:
A separate PR works for me; maybe file an issue to track?